public/app/plugins/datasource/dashboard/datasource.ts

```diff
@@ -111,17 +125,159 @@ export class DashboardDatasource extends DataSourceApi<DashboardQuery> {
           ...s,
           fields: s.fields.map((field: Field) => ({
             ...field,
+            config: {
+              ...field.config,
+              // Enable AdHoc filtering for string and numeric fields only when feature toggle is enabled
+              filterable: config.featureToggles.dashboardDsAdHocFiltering
+                ? field.type === FieldType.string || field.type === FieldType.number
+                : field.config.filterable,
+            },
             state: {
               ...field.state,
             },
           })),
         };
       });
 
-      return [...series, ...annotations];
+      // Apply AdHoc filters to series data (copied and simplified from filterByValue.ts)
+      const filteredSeries =
+        filters.length > 0 ? series.map((frame) => this.applyAdHocFilters(frame, filters)) : series;
+
+      return [...filteredSeries, ...annotations];
+    }
+  }
+
+  /**
+   * Apply AdHoc filters to a DataFrame
+   * Optimized version with pre-computed field indices and value matchers for better performance
+   */
+  private applyAdHocFilters(frame: DataFrame, filters: AdHocVariableFilter[]): DataFrame {
+    if (filters.length === 0 || frame.length === 0) {
+      return frame;
+    }
+
+    // Pre-compute field indices and value matchers for better performance
+    const filterFieldIndices = filters
+      .map((filter) => {
+        const fieldIndex = frame.fields.findIndex((f) => f.name === filter.key);
+        return { filter, fieldIndex, matcher: this.createValueMatcher(filter, fieldIndex, frame) };
+      })
+      .filter(({ filter, fieldIndex, matcher }) => {
+        // If field is not present:
+        // - Keep filters with '=' operator (will always be false - reject rows)
+        // - Remove filters with '!=' operator (will always be true - no effect)
+        if (fieldIndex === -1) {
+          return filter.operator === '=';
+        }
+        // Only keep filters with valid matchers
+        return matcher !== null;
+      });
+
+    // If no filters remain after optimization, return original frame
+    if (filterFieldIndices.length === 0) {
+      return frame;
+    }
+
+    // Short-circuit: if any filter has '=' operator with missing field, reject all rows
+    const hasImpossibleFilter = filterFieldIndices.some(({ fieldIndex }) => fieldIndex === -1);
+    if (hasImpossibleFilter) {
+      return this.reconstructDataFrame(frame, new Set<number>());
```

Comment:
I think the second param here can be optional in the definition? and in this call you could only send frame?

Reply:
This is definitely a nit, I was just trying to avoid instantiating the unused set, but I don't have a strong preference.